### PR TITLE
feature/W-12285553 add retryable messages to retry

### PR DESCRIPTION
**Pull request by barrerajuanjose**
# Description
Add retryable messages option thus it is easier to retry [this kind of errors](https://www.google.com/url?q=https://trailblazer.salesforce.com/issues_view?id%3Da1p4V000002e0HyQAI%26title%3Dintermittent-unexpected-internal-servlet-state-during-soap-and-bulk-api-requests&sa=D&source=editors&ust=1676080657200528&usg=AOvVaw3VoSUUa1W5icHvaJio4Z_w).

# Resolves

[W-12285553](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001IOeF2YAL/view)

# Testing

Here you place all the testing you completed to ensure that the issue was corrected.

## Tests in the PR

- [X] Manual Tests
- [X] Unit Tests
- [X] Integration Tests

### CODE REVIEW

**Review by jaizpuru-sf — CHANGES_REQUESTED** [review]
Thanks for taking care of this JJ! I have a couple of comments:
- Have you considered extending this so that it also covers Partner API? The Exceptions seems to differ across services, according to the document you linked.
- This solution seems to rely heavily on the error message itself, which is potentially subject to change. Some alternatives I can think of: 
    - Rely on the AsyncApiException.exceptionCode (at least for BulkApi), which we already handle in the RetryTemplate.
    - Consider covering the whole class (i.e. SoapFaultException)
    - Keep the message matching approach, but use shorter substrings of the actual messages (i.e.: "Unexpected internal servlet state").

**Comment by barrerajuanjose** [comment]
That makes sense to me @jaizpuru-sf .
I will create a WI about it. Do you think that in that WI we should move this retry logic to salesforce-jdbc repo?

**Comment by ztaylor797** [comment]
> Is it required for 2.5.0? If this error we're seeing is because of some sort of regression, then yes, of course. If not, probably 2.6

Agree with this logic.

> About moving to salesforce-jdbc: I think we'd be better of using stock wsc (or as close as we can get). But I'm also not sure if/how we could move all the retry logic downstream without adding much more overall code volume, which we also don't want.

Stock would be a bit difficult with the interceptors for the arbiter. Not sure how we would handle those.

**Review by ztaylor797 — CHANGES_REQUESTED** [review]
LGTM other than existing comments from Juan.

File: src/main/java/com/sforce/async/BulkConnection.java [hunk 1]
```diff
@@ -196,7 +201,17 @@ static void parseAndThrowException(InputStream in, ContentType type) throws Asyn
...
-                throw new AsyncApiException("Server error returned in unknown format", AsyncExceptionCode.ClientInputError);
+                String body = "?";
+                String typeDesc = type == null ? "?" : type.name();
+                try {
+                    body = new String(FileUtil.toBytes(in), StandardCharsets.UTF_8);
+                    body = body.length() <= 250 ? body : body.substring(0, 250) + "...";
+                } catch (Exception ignored) {
+                }
+                throw new AsyncApiException(
+                        "Server error returned in unknown format " + typeDesc + ". Body: " + body,
```
[the remark this answers is not in the document]
Comment (jaizpuru-sf) [reply src/main/java/com/sforce/async/BulkConnection.java:212]: Makes sense! Addressed

File: src/main/java/com/sforce/async/BulkConnection.java [hunk 2]
```diff
@@ -196,7 +201,17 @@ static void parseAndThrowException(InputStream in, ContentType type) throws Asyn
                 JsonParser parser = factory.createJsonParser(in);
                 exception = parser.readValueAs(AsyncApiException.class);
             } else {
-                throw new AsyncApiException("Server error returned in unknown format", AsyncExceptionCode.ClientInputError);
+                String body = "?";
+                String typeDesc = type == null ? "?" : type.name();
+                try {
+                    body = new String(FileUtil.toBytes(in), StandardCharsets.UTF_8);
+                    body = body.length() <= 250 ? body : body.substring(0, 250) + "...";
```
[the remark this answers is not in the document]
Comment (jaizpuru-sf) [reply src/main/java/com/sforce/async/BulkConnection.java:208]: The thing is, wsc doesn't depend on Apache commons-lang, and I couldn't find anything equivalent

**Comment by ztaylor797** [comment]
That's fine then. Only other thing I might mention is that there is a
single character ellipsis if you wanted to use that.

**Pull request merged**